### PR TITLE
WIP: Introduce new PosLenString/PosLenStringVector types

### PR DESCRIPTION
**Pull request by quinnj**
The aim of these new types is to generalize an idea started in CSV.jl
where we wish to 1) avoid allocating full Strings while parsing and 2)
keep the "string" data as lazy as possible post-parsing. We move the
code for `PosLen` directly from CSV.jl, which is basically a `UInt64`
that stores 20 bits for a strings length, 42 bits for the byte position
where a string begins in a data `Vector{UInt8}` buffer, 1 bit to flag if
a string value is actually `missing`, and the last 1 bit to flag if the
string value has escape characters and needs to be unescaped.

The PosLenStringVector type takes a data buffer, be it a mmapped file or
otherwise, and a `Vector{PosLen}`, along with the escape character (as a
UInt8). When indexed, this custom array type will return a
`PosLenString`, which keeps a reference to the original data buffer (yet
may still be stack allocated due to new compiler optimizations), along
with the string's specific `PosLen` to know where its data lives. Note
that once a `PosLenString` has been materialized, we no longer care
about the `missing` bit, since that must be handled in the
`PosLenStringVector` getindex method.

The `PosLenString` type is an `AbstractString`, and I think it has
really good coverage and performance compared to `Base.String`. I
manually checked performance of each method tested in the
`test/poslenstrings.jl` file. The two cases where there's a slight
performance hit (30-50% slowdown) are `lpad`/`rpad` and `replace`, where
Base algorithms currently force conversion to `String`, which incurs an
extra copy for `PosLenString`. I'll investigate whether we can just copy
the algorithms for those.

Note that mush of the `PosLenString` code was copied from Base routines
due to the current code organization which has optimizations for
`String`, but slower fallbacks for `AbstractString`. `PosLenString`
operates very similarly to `String`, so all those optimizations make
sense for it as well, hence the large amount of copy/paste.

Marking this PR as WIP because I've hardly spent any time on
`PosLenStringVector` yet and need to spend the proper time ensuring it
works properly and considering how we want to support mutability (either
requiring user to do explicit `copy` or some kind of copy-on-write
automatically on first mutating operation). The expectation is that when
users request `lazystrings=true` from CSV.read, they'll be returned a
`PosLenStringVector`, and may need to mutate values afterwards, and I'd
like to support that if at all possible.

### CODE REVIEW

File: src/poslenstrings.jl [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+Base.convert(::Type{UInt64}, x::PosLen) = UInt64(x)
+
+const MISSING_BIT = 0x8000000000000000
+missingvalue(x) = (UInt64(x) & MISSING_BIT) == MISSING_BIT
+
+const ESCAPE_BIT = 0x4000000000000000
+escapedvalue(x) = (UInt64(x) & ESCAPE_BIT) == ESCAPE_BIT
+
+getpos(x) = (UInt64(x) & 0x3ffffffffff00000) >> 20
+getlen(x) = UInt64(x) & 0x00000000000fffff
```
Comment (bkamins) [inline src/poslenstrings.jl:15]: This limit on string size seems OK, but what happens when it is exceeded?

File: src/poslenstrings.jl [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+Base.convert(::Type{PosLen}, x::UInt64) = PosLen(x)
+Base.convert(::Type{UInt64}, x::PosLen) = UInt64(x)
+
+const MISSING_BIT = 0x8000000000000000
+missingvalue(x) = (UInt64(x) & MISSING_BIT) == MISSING_BIT
+
+const ESCAPE_BIT = 0x4000000000000000
+escapedvalue(x) = (UInt64(x) & ESCAPE_BIT) == ESCAPE_BIT
+
+getpos(x) = (UInt64(x) & 0x3ffffffffff00000) >> 20
```
Comment (bkamins) [inline src/poslenstrings.jl:14]: So this is 4TB limit on the file size. Right? Seems OK, but as below - what do we do when it is exceeded?

File: src/poslenstrings.jl [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+getlen(x) = UInt64(x) & 0x00000000000fffff
+
+@noinline lentoolong(len) = throw(ArgumentError("len = $len too long"))
+
+@inline function PosLen(pos, len, missing=false, escaped=false)
+    len > 1048575 && lentoolong(len)
```
Comment (bkamins) [inline src/poslenstrings.jl:20]: OK. I see what happens :). Maybe add type annotations to `PosLen` arguments?

File: src/poslenstrings.jl [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+
+Custom array 
+"""
+struct PosLenStringVector{T} <: AbstractVector{T}
+    data::Vector{UInt8}
+    poslens::Vector{PosLen}
+    e::UInt8
+end
+
+# convenience constructor for testing
```
Comment (bkamins) [inline src/poslenstrings.jl:456]: maybe move these convenience constructors to tests?

File: src/poslenstrings.jl [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+# copy:
+  # calcs total len of all strings in poslens
+  # allocates fresh data byte buffer for total len
+  # loops through and copies each strings data into new buffer, recalcs poslen
+"""
+    PosLenStringVector{T}(data, poslens, e=UInt8('\\'))
+
+Custom array 
+"""
+struct PosLenStringVector{T} <: AbstractVector{T}
```
Comment (bkamins) [inline src/poslenstrings.jl:450]: why does this type have a free parameter `T`?

File: src/poslenstrings.jl (hunk 2 above)
Comment (quinnj) [reply src/poslenstrings.jl:14]: We can/should add an error check in CSV.jl if an input file happens to be > 4TB and they request `lazystrings=true`.

File: src/poslenstrings.jl (hunk 3 above)
Comment (quinnj) [reply src/poslenstrings.jl:20]: type annotations? Do you mean code comments explaining things more?

Comment (quinnj) [reply src/poslenstrings.jl:20]: oh, I see what you mean. yeah, I can do that.

File: src/poslenstrings.jl (hunk 5 above)
Comment (quinnj) [reply src/poslenstrings.jl:450]: This allows callers/users to specify whether the element type is `Union{Missing, PosLenString}` or just plain `PosLenString`. in CSV.jl, we keep track if any `missing` values were detected while parsing, so we can set element type to just `PosLenString` for better type stability when calling `getindex`

File: src/poslenstrings.jl (hunk 2 above)
Comment (quinnj) [reply src/poslenstrings.jl:14]: Ok, added check in CSV.jl here: https://github.com/JuliaData/CSV.jl/pull/837/commits/c52b8b36d27c24057ed303174991fe367eb35290

**Pull request closed without merging**